### PR TITLE
Restructure project for tools support core repository

### PR DESCRIPTION
**Pull request by BenchR267**
This renames Basic, Utility and their dependencies to TSCBasic and TSCUtility.
This makes it possible to copy their directories to the tools-support-core repository to be able to use them in other projects too.
TestSupport is now divided in TSCTestSupport and SPMTestSupport where the latter contains all SwiftPM specific parts and TSC the generic parts which use the tools support core interfaces.

The idea is to copy the sources and tests to the tools-support-core repository and eventually remove it from here and depend on the new repo. That requires some support in cmake which we don't rely on yet. More information about the motivation can be read here: https://forums.swift.org/t/factoring-out-a-new-repo-for-shared-code-between-llbuild-swiftpm/11429/22

Related PRs:
https://github.com/apple/sourcekit-lsp/pull/153
https://github.com/apple/swift-tools-support-core/pull/1

rdar://55022772

### CODE REVIEW

**Comment by BenchR267** [comment]
@swift-ci please smoke test

**Comment by benlangmuir** [comment]
Once you figure out the bootstrap issues you'll need to test this with an updated sourcekit-lsp as well.  I created https://github.com/apple/sourcekit-lsp/pull/153 with the module name changes.

Incidentally, I noticed while updating the code that there are some APIs named `spm_*` such as `spm_chomp` and `spm_only` that should probably be renamed.

File: Documentation/PackageDescription.md [hunk 1]
```diff
@@ -21,10 +21,10 @@ let package = Package(
         .library(name: "MyLibrary", targets: ["MyLibrary"]),
     ],
     dependencies: [
-        .package(url: "https://url/of/another/package/named/Utility", from: "1.0.0"),
+        .package(url: "https://url/of/another/package/named/TSCUtility", from: "1.0.0"),
```
[the remark this answers is not in the document]
Comment (BenchR267) [reply Documentation/PackageDescription.md]: Oh sure, I thought I checked all occurrences of false positive renames 🙂

**Comment by BenchR267** [comment]
@swift-ci please smoke test

File: Utilities/PerformanceTests.xcscheme [hunk 2]
```diff
@@ -17,8 +17,8 @@
             <BuildableReference
                BuildableIdentifier = "primary"
                BlueprintIdentifier = "OBJ_638"
-               BuildableName = "BasicPerformanceTests.xctest"
-               BlueprintName = "BasicPerformanceTests"
+               BuildableName = "TSCBasicPerformanceTests.xctest"
+               BlueprintName = "TSCBasicPerformanceTests"
```
[the remark this answers is not in the document]
Comment (BenchR267) [reply Utilities/PerformanceTests.xcscheme]: Yep, I copied the scheme-file into the project's shared data and it picks it up correctly and finds the buildable reference 👍

**Comment by BenchR267** [comment]
@swift-ci please smoke test

**Comment by BenchR267** [comment]
I think I might need to factor more stuff out that is interesting for other projects. For example the ArgumentParser is part of TSCUtility but the whole Commands package defines the tool infrastructure which we might want to refactor out to TSCUtility a bit.
What do you think @aciidb0mb3r? Nevertheless I think we can iterate on the changes in this PR later on after having the first version landed.

**Comment by BenchR267** [comment]
@swift-ci please smoke test

**Comment by BenchR267** [comment]
@benlangmuir I tested your changes in SourceKit-lsp with the changes here and they worked fine (thanks for working on them!). How can I configure the CI to use your branch instead of master here?

**Comment by benlangmuir** [comment]
Here are the details: https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#cross-repository-testing

**Comment by benlangmuir** [comment]
https://github.com/apple/sourcekit-lsp/pull/153

@swift-ci please smoke test

**Comment by BenchR267** [comment]
Thanks @benlangmuir! It didn't work though. It's complaining about `Sources/SKCore/BuildServerBuildSystem.swift:12:8: error: no such module 'Basic'` but that file doesn't exist in your branch. Does master needs to be rebased into your branch?

**Comment by benlangmuir** [comment]
apple/sourcekit-lsp#153

@swift-ci please smoke test

**Comment by aciidb0mb3r** [comment]
> I think I might need to factor more stuff out that is interesting for other projects. For example the ArgumentParser is part of TSCUtility but the whole Commands package defines the tool infrastructure which we might want to refactor out to TSCUtility a bit.

That one is a bit SwiftPM specific that doesn't necessarily make sense as a generic utility but we can discuss that.

What do you think @aciidb0mb3r? Nevertheless I think we can iterate on the changes in this PR later on after having the first version landed.

Yep, let's land with minimal changes first and then we refactor incrementally.

**Comment by BenchR267** [comment]
Yeah, I don't mean to move SwiftTool to TSC but more provide infrastructure that SwiftTool can use as well as others. Basically functionality for providing usage information and handling the argument parser calling.

**Comment by BenchR267** [comment]
apple/sourcekit-lsp#153

@swift-ci please smoke test

**Comment by BenchR267** [comment]
apple/sourcekit-lsp#153

@swift-ci please smoke test

**Comment by BenchR267** [comment]
apple/sourcekit-lsp#153

@swift-ci please smoke test

**Pull request merged**